File: Others/prj_lowe_1.cpp
/*::::::::::::::::::::::::::::::::::::::::::::::::::::*/
/*:::::::::::::J A I  S H R E E  R A M::::::::::::::::*/
/*::::::::::::::::::::::::::::::::::::::::::::::::::::*/

#include <bits/stdc++.h>
using namespace std;
#ifdef LUCTIVUD
    #include <buggyBaby.hpp>
    pretty:: PrettyPrinter NonIterable;
    #define cerr cout
#else
    #define _____error_____(...)
#endif
#pragma GCC optimize "trapv"

/*:::::::::::::::::::SNIPS::::::::::::::::::::::::::::*/

typedef long long lld; typedef unsigned long long llu;

#define  forn(I7, E4) for(lld I7=0ll; I7 < E4; (I7)+=1ll)
#define forn1(I7, E4) for(lld I7=1ll; I7 < E4+1; (I7)+=1ll)
#define        len(v) ((int)((v).size()))
#define        all(x) (x).begin(), (x).end()
#define       rall(x) (x).rbegin(), (x).rend()
#define            f1 first
#define            s2 second

/*::::::::::::::::::/SNIPS::::::::::::::::::::::::::::*/
/*:::::::::::::::::::UTILS::::::::::::::::::::::::::::*/

const long double EPS = 1e-6;
lld MOD = int(1e9) + 7;
lld &mod = MOD;






/*:::::::::::::::::::::/UTILS:::::::::::::::::::::::::*/
/*::::::::::::::::::::::LOGIC:::::::::::::::::::::::::*/



void awesomeSort(int n, vector<int> &arr) {
    // cout << "Case #" << _TestCase << ": ";
    vector<pair<int, int>> arr2(n);
    for (int i = 0; i < n; i++) {
        arr2[i] = {arr[i], i};
    }

    sort(arr2.begin(), arr2.end(), [] (pair<int, int> &A, pair<int, int> &B) {
        int a = A.first, b = B.first;
        int aPos = A.second, bPos = B.second;
        if ((a & 1) == (b & 1)) {
            if (a & 1) { // both odd
                return aPos < bPos;
            } else { // both even
                if (a % 5 == 0 and b % 5 == 0) {
                    return a > b;
                } else if (a % 5 == 0) {
                    return true;
                } else if (b % 5 == 0) {
                    return false;
                } else {
                    return aPos > bPos;
                }
            }
        } else {
            return (a & 1) < (b & 1);
        }
    });

    for (int i = 0; i < n; i++) {
        arr[i] = arr2[i].first;
    }
}



/*:::::::::::::::::::::/LOGIC:::::::::::::::::::::::::*/
/*::::::::::::::::::::::INVOC:::::::::::::::::::::::::*/
signed main() {
    ios_base::sync_with_stdio(false); cin.tie(0);
    // cout.precision(10); cout << fixed;
    #ifdef LUCTIVUD
      // const auto start_time = std::chrono::high_resolution_clock::now();
      freopen("/home/luctivud/CPPractice/Zinput.txt", "r", stdin);
      freopen("/home/luctivud/CPPractice/Zoutput.txt", "w", stdout);
    #endif


    MOD = mod = lld(1e9) + 7ll;
    int _T0T4 = 1;
    cin >> _T0T4; 

    for (int _TestCase = 1; _TestCase <= _T0T4; _TestCase++) {
        int n; cin >> n;
        vector<int> arr(n);
        forn(i, n) {
            cin >> arr[i];
        }
        awesomeSort(n, arr);
        // _____error_____(ans);
        _____error_____(arr);
        cout << "\n"; 
    }


    #ifdef LUCTIVUD
      // auto end_time = std::chrono::high_resolution_clock::now();
      // std::chrono::duration<double> diff = end_time - start_time;
      // cerr << "Finished in : " << diff.count() << "\n";
    #endif

    return 0; 
}


/*:::::::::::::::::::::/INVOC:::::::::::::::::::::::::*/
/*  ~~
Author   :  Udit "luctivud" Gupta 
linkedin :  https://www.linkedin.com/in/udit-gupta-1b7863135/
*/

File: Others/prj_lowe_1.py
#   J A I  S H R E E  R A M   #


import math, sys, collections, functools, time, itertools;

# sys.setrecursionlimit(10**6)


def Read_Ints()     : return map(int, input().strip().split())
def Read_Array()    : return list(Read_Ints())
def Read_Strings()  : return list(input().strip().split())
def printxsp(*args) : return print(*args, end="")
def printsp(*args)  : return print(*args, end=" ")



d4i = [-1, +0, +1, +0]; d8i = [-1, -1, +0, +1, +1, +1, +0, -1]; 
d4j = [+0, +1, +0, -1]; d8j = [+0, +1, +1, +1, +0, -1, -1, -1];


# >>>>>>--->>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>





def awesomeSort(n, arr):
    # printsp("Case #{}: ".format(_TestCase)) 
    for i, v in enumerate(arr):
        arr[i] = [arr[i], i]
    arr.sort(key = lambda x : ((x[0] & 1, x[0] % 5, -x[0], -x[1]), x[1]))
    return arr
    






_T0T4 = 1;
_T0T4 = int(input()) 
for _TestCase in range(1, _T0T4 + 1): 
    # solveEachTest(_TestCase)
    n = int(input())
    arr = Read_Array()
    print(awesomeSort(n, arr))


# Udit "luctivud" Gupta 
# linkedin :  https://www.linkedin.com/in/udit-gupta-1b7863135/

